x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/event/AnalyticsEvent.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.application.analytics.event;
+
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentType;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * This class represents Analytics events object meant to be emitted to the event queue.
+ *
+ * An analytics events is composed of:
+ * - A context ({@link AnalyticsContext}) containing basic event meta-data (type, time, collection)
+ * - A payload containing a JSON representation of the event.
+ *
+ * The payload has to be validated. The best way to do it is to create events using the {@link AnalyticsEventParser}.
+ */
+public class AnalyticsEvent implements Writeable, ToXContentObject {
+    private final AnalyticsContext analyticsContext;
+
+    private final XContentType xContentType;
+
+    private final BytesReference payload;
+
+    public AnalyticsEvent(AnalyticsContext analyticsContext, XContentType xContentType, BytesReference payload) {
+        this.analyticsContext = Objects.requireNonNull(analyticsContext, "analyticsContext");
+        this.xContentType = Objects.requireNonNull(xContentType, "xContentType");
+        this.payload = Objects.requireNonNull(payload, "payload");
+    }
+
+    protected AnalyticsEvent(StreamInput in) throws IOException {
+        this(new AnalyticsContext(in), in.readEnum(XContentType.class), in.readBytesReference());
+    }
+
+    public AnalyticsContext getAnalyticsContext() {
+        return analyticsContext;
+    }
+
+    public XContentType getxContentType() {
+        return xContentType;
+    }
+
+    public BytesReference getPayload() {
+        return payload;
+    }
+
+    public Map<String, Object> getPayloadAsMap() {
+        return XContentHelper.convertToMap(payload, true, xContentType).v2();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        analyticsContext.writeTo(out);
+        XContentHelper.writeTo(out, xContentType);
+        out.writeBytesReference(payload);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
```

Comment:
@jimczi left a [comment](https://github.com/elastic/elasticsearch/pull/94404#discussion_r1132406852) in my PR that it's better to not use `toXContent` and replace it with `ESLogMessage` instead

Reply:
I was missing the context in this comment. I think it's fine here since we need to parse the event for validation?